lib/mixins/request.js

```diff
@@ -11,167 +11,96 @@ except according to the terms contained in the LICENSE file.
-  return result;
-};
-
-// Used for consecutive (not concurrent) requests.
-function request(config) {
-  const id = this.requestId != null ? this.requestId + 1 : 0;
-  this.requestId = id;
-  return this.$http.request(config)
-    .catch(onErrorResponse(this, id))
-    .then(onSuccessfulResponseOrResponses(this, id));
-}
-
-// Used for concurrent requests.
-function requestAll(requests) {
-  const id = this.requestId != null ? this.requestId + 1 : 0;
-  this.requestId = id;
-  return Promise.all(requests)
-    .catch(onErrorResponse(this, id))
-    .then(onSuccessfulResponseOrResponses(this, id));
-}
-
-function get(url, config) {
-  return this.request({ ...config, method: 'get', url });
-}
-
-function post(url, data, config) {
-  return this.request({ ...config, method: 'post', url, data });
-}
-
-function put(url, data, config) {
-  return this.request({ ...config, method: 'put', url, data });
-}
-
-function patch(url, data, config) {
-  return this.request({ ...config, method: 'patch', url, data });
-}
-
-function del(url, config) {
-  return this.request({ ...config, method: 'delete', url });
-}
-
-function maybeGet(options) {
-  const entries = Object.entries(options);
-  if (entries.length === 0) throw new Error('invalid options');
-
-  const id = this.requestId != null ? this.requestId + 1 : 0;
-  this.requestId = id;
-
-  let responseCount = 0;
-  let firstError = true;
-  for (const entry of entries) {
-    const [prop, { url, extended = false, transform = R.identity }] = entry;
-    this[prop] = MaybeData.awaiting();
-    const headers = {};
-    if (extended) headers['X-Extended-Metadata'] = 'true';
-    this.$http.get(url, { headers })
-      .catch(error => { // eslint-disable-line no-loop-func
-        throwErrorForIdMismatch(this.requestId, id);
-
-        logRequestError(error);
-        if (firstError) this.$alert().danger(alertMessage(this, error));
-
-        this[prop] = MaybeData.error();
-        responseCount += 1;
-        // We reset this.requestId in the same tick that we set this[prop] so
-        // that this[prop] remains in sync with this.requestId (and
-        // this.awaitingResponse).
-        if (responseCount === entries.length) this.requestId = null;
-
-        firstError = false;
-        throw error;
-      })
-      .then(({ data }) => { // eslint-disable-line no-loop-func
-        throwErrorForIdMismatch(this.requestId, id);
-
-        this[prop] = MaybeData.success(transform(data));
-        responseCount += 1;
-        if (responseCount === entries.length) this.requestId = null;
-      })
-      // This catch will be triggered if there was an axios error, including for
-      // an error response, or if there was a request ID mismatch.
-      .catch(() => {});
-  }
+  const { currentRoute } = this.$store.state.router;
+  return this.$http.request(configForPossibleBackendRequest(axiosConfig, token))
+    .catch(error => {
+      if (this.$store.state.router.currentRoute !== currentRoute)
```

Comment:
I think we're going to keep the current behavior for now.

We do not run the `then()` callback on a `POST`/`PUT`/`PATCH`/`DELETE` request if the user has navigated away from the route, because the callback often completes actions that are only appropriate if the route has not changed — for example, displaying an alert or navigating to another route. I think it would add complexity to divide actions between those that are appropriate only if the route has not changed and those that should be completed even if the route has changed.

I also don't think it's very unexpected that Frontend will not reflect these updates to the data: if a user requests something on a page, then navigates away from the page while the request is in progress, I don't think it's very unexpected that any resulting updates to the data will not be shown until a page refresh.

Implementing #204 may also alleviate this concern a bit, because then data would be preserved for less time in general, reducing the chance of stale or inconsistent data.